bee-tangle/src/ms_tangle.rs

```diff
@@ -0,0 +1,1145 @@
+    }
+
+    /// Shut down the tangle, terminating any and all worker tasks.
+    pub async fn shutdown(self) {
+        // TODO: Write back changes by calling self.inner.shutdown().await
+    }
+
+    /// Get the configuration of this tangle.
+    pub fn config(&self) -> &TangleConfig {
+        &self.config
+    }
+
+    /// Insert a message into the tangle.
+    pub async fn insert(&self, message: Message, hash: MessageId, metadata: MessageMetadata) -> Option<MessageRef> {
+        // TODO this has been temporarily moved to the processor.
+        // Reason is that since the tangle is not a worker, it can't have access to the propagator tx.
+        // When the tangle is made a worker, this should be put back on.
+        // if opt.is_some() {
+        //     if let Err(e) = Protocol::get()
+        //         .propagator_worker
+        //         .unbounded_send(PropagatorWorkerEvent(hash))
+        //     {
+        //         error!("Failed to send hash to propagator: {:?}.", e);
+        //     }
+        // }
+        //
+        // opt
+        self.inner.insert(hash, message, metadata).await
+    }
+
+    /// Add a milestone to the tangle.
+    pub async fn add_milestone(&self, idx: MilestoneIndex, milestone: Milestone) {
+        // TODO: only insert if vacant
+        self.inner
+            .update_metadata(&milestone.message_id(), |metadata| {
+                metadata.flags_mut().set_milestone(true);
+                metadata.set_milestone_index(idx);
+                metadata.set_omrsi(IndexId::new(idx, *milestone.message_id()));
+                metadata.set_ymrsi(IndexId::new(idx, *milestone.message_id()));
+            })
+            .await;
+        self.inner
+            .hooks()
+            .insert_milestone(idx, &milestone)
+            .await
+            .unwrap_or_else(|e| info!("Failed to insert message {:?}", e));
+        // TODO can there be a race condition between 2 ops ?
+        self.milestones.lock().await.insert(idx, milestone);
+    }
+
+    /// Remove a milestone from the tangle.
+    pub async fn remove_milestone(&self, index: MilestoneIndex) {
+        self.milestones.lock().await.remove(&index);
+    }
+
+    async fn pull_milestone(&self, idx: MilestoneIndex) -> Option<MessageId> {
+        if let Some(milestone) = self.inner.hooks().get_milestone(&idx).await.unwrap_or_else(|e| {
+            info!("Failed to insert message {:?}", e);
+            None
+        }) {
+            let message_id = *self
+                .milestones
+                .lock()
+                .await
+                .entry(idx)
+                .or_insert(milestone)
+                .message_id();
+
+            Some(message_id)
+        } else {
+            None
+        }
+    }
+
+    /// Get the milestone from the tangle that corresponds to the given milestone index.
+    pub async fn get_milestone(&self, index: MilestoneIndex) -> Option<Milestone> {
+        self.milestones.lock().await.get(&index).cloned()
```

Comment:
Same here.

Reply:
As mentioned before: this is not a problem because a) the mutex is not exposed and b) there is no way for a caller to hold the lock twice because each method is effectively atomic.